core/shared/src/main/scala/zio/Ref.scala

```diff
@@ -379,50 +374,29 @@ object Ref extends Serializable {
     }
 
     val unsafe: UnsafeAPI =
-      new UnsafeAPI {
+      new AtomicReference[A](initial) with UnsafeAPI { ref: AtomicReference[A] =>
         def get(implicit unsafe: Unsafe): A =
-          value.get
+          ref.asInstanceOf[AtomicReference[A]].get
 
-        def getAndSet(a: A)(implicit unsafe: Unsafe): A = {
-          var loop       = true
-          var current: A = null.asInstanceOf[A]
-          while (loop) {
-            current = value.get
-            loop = !value.compareAndSet(current, a)
-          }
-          current
-        }
+        def getAndSet(a: A)(implicit unsafe: Unsafe): A =
+          ref.asInstanceOf[AtomicReference[A]].getAndSet(a)
 
-        def getAndUpdate(f: A => A)(implicit unsafe: Unsafe): A = {
-          var loop       = true
-          var current: A = null.asInstanceOf[A]
-          while (loop) {
-            current = value.get
-            val next = f(current)
-            loop = !value.compareAndSet(current, next)
-          }
-          current
-        }
+        def getAndUpdate(f: A => A)(implicit unsafe: Unsafe): A =
+          ref.asInstanceOf[AtomicReference[A]].getAndUpdate(f.apply)
 
-        def getAndUpdateSome(pf: PartialFunction[A, A])(implicit unsafe: Unsafe): A = {
-          var loop       = true
-          var current: A = null.asInstanceOf[A]
-          while (loop) {
-            current = value.get
-            val next = pf.applyOrElse(current, (_: A) => current)
-            loop = !value.compareAndSet(current, next)
-          }
-          current
-        }
+        def getAndUpdateSome(pf: PartialFunction[A, A])(implicit unsafe: Unsafe): A =
+          ref
+            .asInstanceOf[AtomicReference[A]]
+            .getAndUpdate((current: A) => pf.applyOrElse(current, (_: Any) => current))
 
         def modify[B](f: A => (B, A))(implicit unsafe: Unsafe): B = {
           var loop = true
           var b: B = null.asInstanceOf[B]
           while (loop) {
-            val current = value.get
+            val current = ref.asInstanceOf[AtomicReference[A]].get
```

Comment:
could this whole fn be written as:
```
def modify(f) =
  var b = null
  ref.updateAndGet: a =>
    tup = f(a)
    b = tup._1
    tup._2
  b
```
?

Reply:
This implementation is more efficient. In fact one would argue the most efficient would be to use `weakCompareAndSet` but I think we can leave that one as is for now